File: server/internal/accounts/db/evm_wallets.go
package accountsdb

import (
	"context"
	"database/sql"
	"fmt"
	"strings"
	"time"

	"github.com/samber/lo"

	"tallyo/internal/accounts"
	"tallyo/internal/database/dbgen"
	"tallyo/internal/database/dbutil"
	"tallyo/internal/graph/model"
	u "tallyo/internal/utils"
)

func (s *Store) CreateEVMWallet(
	ctx context.Context,
	address string,
	ownerID int64,
	label string,
	chainIDs []string,
) (*model.Connection, *model.Account, error) {
	owner, err := s.OwnerByID(ctx, ownerID)
	if err != nil {
		return nil, nil, fmt.Errorf("lookup owner: %w", err)
	}
	if owner == nil {
		return nil, nil, fmt.Errorf("unknown owner id %d", ownerID)
	}

	name := label
	if name == "" {
		name = truncateAddress(address)
	}

	accountID, err := dbutil.RandomID()
	if err != nil {
		return nil, nil, err
	}
	var connID int64
	var accountRowID int64
	var walletID int64

	if err := s.WithTx(ctx, func(_ *sql.Tx, q *dbgen.Queries) error {
		newWalletID, err := q.InsertEVMWallet(ctx, dbgen.InsertEVMWalletParams{
			Address:  address,
			Label:    label,
			ChainIds: strings.Join(accounts.NormalizeEVMChainIDs(chainIDs), ","),
		})
		if err != nil {
			return fmt.Errorf("insert evm_wallet: %w", err)
		}
		walletID = newWalletID
		newConnID, err := q.UpsertConnection(ctx, dbgen.UpsertConnectionParams{
			SourceTable: string(accounts.SourceTableEVMWallet),
			SourceID:    walletID,
			Name:        &name,
			OwnerID:     ownerID,
		})
		if err != nil {
			return fmt.Errorf("insert connection: %w", err)
		}
		connID = newConnID
		newAccountID, err := q.InsertLinkedAccount(ctx, dbgen.InsertLinkedAccountParams{
			ExternalID:   accountID,
			ConnectionID: &connID,
			OwnerID:      ownerID,
			Name:         name,
			Type:         string(model.AccountTypeCryptoWallet),
		})
		if err != nil {
			return fmt.Errorf("insert evm account: %w", err)
		}
		accountRowID = newAccountID
		return nil
	}); err != nil {
		return nil, nil, err
	}

	conn := &model.Connection{ID: model.New(model.GlobalIDConnection, connID), Name: &name, Owner: owner, IsActive: true, SourceTable: string(accounts.SourceTableEVMWallet), SourceID: walletID}
	account := &model.Account{
		ID:         model.New(model.GlobalIDAccount, accountRowID),
		Connection: conn,
		Owner:      owner,
		Name:       name,
		Type:       model.AccountTypeCryptoWallet,
		Manual:     false,
	}
	return conn, account, nil
}

func (s *Store) EVMWalletByConnectionID(ctx context.Context, connectionID int64) (*accounts.EVMWallet, error) {
	walletsByConnectionID, err := s.EVMWalletsByConnectionIDs(ctx, []int64{connectionID})
	return dbutil.MapSingle(walletsByConnectionID, err, connectionID)
}

func (s *Store) EVMWalletsByConnectionIDs(ctx context.Context, connectionIDs []int64) (map[int64]*accounts.EVMWallet, error) {
	if len(connectionIDs) == 0 {
		return map[int64]*accounts.EVMWallet{}, nil
	}
	rows, err := s.q.EVMWallets(ctx, dbgen.EVMWalletsParams{ConnectionIds: connectionIDs})
	if err != nil {
		return nil, fmt.Errorf("lookup evm wallets by connection: %w", err)
	}
	toWalletByConnectionID := func(row dbgen.EVMWalletsRow) (int64, *accounts.EVMWallet) {
		wallet := evmWalletFromSQLRow(row)
		return row.ConnectionID, &wallet
	}
	return lo.Associate(rows, toWalletByConnectionID), nil
}

func (s *Store) EVMWalletsDueForBalanceSync(ctx context.Context, now time.Time) ([]accounts.EVMWallet, error) {
	nowUTC := now.UTC()
	rows, err := s.q.EVMWallets(ctx, dbgen.EVMWalletsParams{Now: &nowUTC, DueOnly: true})
	if err != nil {
		return nil, fmt.Errorf("list evm wallets due for balance sync: %w", err)
	}
	return u.Map(rows, evmWalletFromSQLRow), nil
}

func (s *Store) SetEVMWalletBalanceSynced(ctx context.Context, walletID int64, next time.Time) error {
	nextUTC := next.UTC()
	return s.q.SetEVMWalletBalanceSynced(ctx, dbgen.SetEVMWalletBalanceSyncedParams{
		ID:                walletID,
		NextBalanceSyncAt: &nextUTC,
	})
}

func (s *Store) EVMWalletAccountID(ctx context.Context, walletID int64) (int64, error) {
	accountID, err := s.q.EVMWalletAccountID(ctx, dbgen.EVMWalletAccountIDParams{WalletID: walletID})
	identity := func(id int64) int64 { return id }
	return dbutil.MapRow(accountID, err, identity)
}

func (s *Store) DeleteEVMWallet(ctx context.Context, connectionID int64) error {
	wallet, err := s.EVMWalletByConnectionID(ctx, connectionID)
	if err != nil {
		return err
	}
	if wallet == nil {
		return fmt.Errorf("evm wallet for connection %d not found", connectionID)
	}

	return s.WithTx(ctx, func(_ *sql.Tx, q *dbgen.Queries) error {
		if err := q.DeleteAccountsByConnection(ctx, dbgen.DeleteAccountsByConnectionParams{ConnectionID: &connectionID}); err != nil {
			return fmt.Errorf("delete accounts: %w", err)
		}
		if err := q.DeleteConnectionByID(ctx, dbgen.DeleteConnectionByIDParams{ID: connectionID}); err != nil {
			return fmt.Errorf("delete connection: %w", err)
		}
		if err := q.DeleteEVMWallet(ctx, dbgen.DeleteEVMWalletParams{ID: wallet.ID}); err != nil {
			return fmt.Errorf("delete evm_wallet: %w", err)
		}
		return nil
	})
}

func evmWalletFromSQLRow(row dbgen.EVMWalletsRow) accounts.EVMWallet {
	return accounts.EVMWallet{
		ID:                row.ID,
		Address:           row.Address,
		ChainIDs:          dbutil.SplitCSV(row.ChainIds),
		OwnerID:           row.OwnerID,
		Label:             row.Label,
		CreatedAt:         row.CreatedAt,
		NextBalanceSyncAt: row.NextBalanceSyncAt,
	}
}

func truncateAddress(addr string) string {
	if len(addr) <= 10 {
		return addr
	}
	return addr[:6] + "…" + addr[len(addr)-4:]
}

File: server-rs/src/accounts/store/evm_wallets.rs
use std::collections::HashMap;

use anyhow::{Context, Result, anyhow};
use chrono::{DateTime, Utc};
use rand::RngCore;
use sqlx::{Executor, Sqlite, SqlitePool};

use crate::{
    accounts::{Account, AccountType, ConnectionRecord, EvmWallet, SourceTable, normalize_evm_chain_ids},
    database::{self, queries},
};

use super::{account_by_id, connection_by_id, owner_by_id};

pub async fn create_evm_wallet(
    pool: &SqlitePool,
    address: &str,
    owner_id: i64,
    label: &str,
    chain_ids: &[String],
) -> Result<(ConnectionRecord, Account)> {
    let address = address.to_owned();
    if owner_by_id(pool, owner_id).await.context("lookup owner")?.is_none() {
        return Err(anyhow!("unknown owner id {owner_id}"));
    }
    let name = if label.is_empty() { truncate_address(&address) } else { label.to_owned() };
    let wallet_label = label.to_owned();
    let chain_ids = normalize_evm_chain_ids(chain_ids).join(",");
    let external_id = random_id();
    let stored_name = name.clone();
    let source_table = SourceTable::EvmWallets.to_string();
    let account_type = AccountType::CryptoWallet.to_string();
    let (_wallet_id, connection_id, account_id) = database::with_tx(pool, |transaction| {
        Box::pin(async move {
            let wallet_id = queries::insert_evm_wallet(
                &mut **transaction,
                queries::InsertEvmWalletParams {
                    address: &address,
                    label: &wallet_label,
                    chain_ids: &chain_ids,
                },
            )
            .await
            .context("insert evm_wallet")?
            .id;
            let connection_id = queries::upsert_connection(
                &mut **transaction,
                queries::UpsertConnectionParams {
                    source_table: &source_table,
                    source_id: wallet_id,
                    name: Some(&stored_name),
                    owner_id,
                },
            )
            .await
            .context("insert connection")?
            .id;
            let account_id = queries::insert_linked_account(
                &mut **transaction,
                queries::InsertLinkedAccountParams {
                    external_id: &external_id,
                    connection_id: Some(connection_id),
                    owner_id,
                    name: &stored_name,
                    r#type: &account_type,
                },
            )
            .await
            .context("insert evm account")?
            .id;
            Ok((wallet_id, connection_id, account_id))
        })
    })
    .await?;
    let connection = connection_by_id(pool, connection_id)
        .await?
        .ok_or_else(|| anyhow!("connection {connection_id} not found"))?;
    let account = account_by_id(pool, account_id)
        .await?
        .ok_or_else(|| anyhow!("account {account_id} not found"))?;
    Ok((connection, account))
}

pub async fn evm_wallet_by_connection_id(pool: &SqlitePool, connection_id: i64) -> Result<Option<EvmWallet>> {
    let wallets = evm_wallets_by_connection_ids(pool, &[connection_id]).await?;
    Ok(wallets.into_values().next())
}

pub async fn evm_wallets_by_connection_ids(
    executor: impl Executor<'_, Database = Sqlite>,
    connection_ids: &[i64],
) -> Result<HashMap<i64, EvmWallet>> {
    if connection_ids.is_empty() {
        return Ok(HashMap::new());
    }
    queries::evm_wallets(
        executor,
        queries::EvmWalletsParams {
            connection_ids: Some(connection_ids),
            ..Default::default()
        },
    )
    .await
    .map(|rows| {
        rows.into_iter()
            .map(|row| {
                let connection_id = row.connection_id;
                (connection_id, row.into())
            })
            .collect()
    })
    .context("lookup evm wallets by connection")
}

pub async fn evm_wallets_due_for_balance_sync(
    executor: impl Executor<'_, Database = Sqlite>,
    now: DateTime<Utc>,
) -> Result<Vec<EvmWallet>> {
    queries::evm_wallets(
        executor,
        queries::EvmWalletsParams {
            now: Some(now.into()),
            due_only: true,
            ..Default::default()
        },
    )
    .await
    .map(|rows| rows.into_iter().map(Into::into).collect())
    .context("list evm wallets due for balance sync")
}

pub async fn set_evm_wallet_balance_synced(
    executor: impl Executor<'_, Database = Sqlite>,
    wallet_id: i64,
    next_balance_sync_at: DateTime<Utc>,
) -> Result<()> {
    queries::set_evm_wallet_balance_synced(
        executor,
        queries::SetEvmWalletBalanceSyncedParams {
            next_balance_sync_at: Some(next_balance_sync_at.into()),
            id: wallet_id,
        },
    )
    .await
    .map_err(Into::into)
}

pub async fn evm_wallet_account_id(
    executor: impl Executor<'_, Database = Sqlite>,
    wallet_id: i64,
) -> Result<Option<i64>> {
    queries::evm_wallet_account_id_opt(executor, queries::EvmWalletAccountIdParams { wallet_id })
        .await
        .map(|row| row.map(|row| row.id))
        .map_err(Into::into)
}

pub async fn delete_evm_wallet(pool: &SqlitePool, connection_id: i64) -> Result<()> {
    let wallet = evm_wallet_by_connection_id(pool, connection_id)
        .await?
        .ok_or_else(|| anyhow!("evm wallet for connection {connection_id} not found"))?;
    database::with_tx(pool, |transaction| {
        Box::pin(async move {
            queries::delete_accounts_by_connection(
                &mut **transaction,
                queries::DeleteAccountsByConnectionParams {
                    connection_id: Some(connection_id),
                },
            )
            .await
            .context("delete accounts")?;
            queries::delete_connection_by_id(
                &mut **transaction,
                queries::DeleteConnectionByIdParams { id: connection_id },
            )
            .await
            .context("delete connection")?;
            queries::delete_evm_wallet(&mut **transaction, queries::DeleteEvmWalletParams { id: wallet.id })
                .await
                .context("delete evm_wallet")?;
            Ok(())
        })
    })
    .await
}

pub fn truncate_address(address: &str) -> String {
    if address.len() <= 10 {
        return address.to_owned();
    }
    format!("{}…{}", &address[..6], &address[address.len() - 4..])
}

fn random_id() -> String {
    let mut bytes = [0; 16];
    rand::rng().fill_bytes(&mut bytes);
    hex::encode(bytes)
}

#[cfg(test)]
mod tests {
    use anyhow::Result;

    use crate::{database::dbtest, testutil::store::create_owner};

    use super::*;

    #[tokio::test]
    async fn evm_wallet_creation_returns_the_persisted_rows() -> Result<()> {
        let pool = dbtest::open().await?;
        let owner = create_owner(&pool, "alex").await?;
        let (connection, account) = create_evm_wallet(
            &pool,
            "0x1111111111111111111111111111111111111111",
            owner.id,
            "",
            &["eth".into()],
        )
        .await?;
        let wallet = evm_wallet_by_connection_id(&pool, connection.connection.id)
            .await?
            .unwrap();
        assert_eq!(wallet.label, "");
        assert_eq!(
            super::super::connection_by_id(&pool, connection.connection.id).await?,
            Some(connection.clone())
        );
        assert_eq!(
            super::super::account_by_id(&pool, account.id).await?,
            Some(account.clone())
        );
        assert_eq!(evm_wallet_account_id(&pool, wallet.id).await?, Some(account.id));
        Ok(())
    }

    #[tokio::test]
    async fn empty_batch_lookups_return_empty_maps() -> Result<()> {
        let pool = dbtest::open().await?;
        assert!(evm_wallets_by_connection_ids(&pool, &[]).await?.is_empty());
        Ok(())
    }
}
